V3.2/task/flavor template

pkumarex opened the pull request:
Includes the FlavorTemplate API changes, swagger documentation.
Note :
*Models in FlavorSchema.go has been included for swagger. It will be removed, once the Flavor creation is added.
*Conditions to check for "AMI/BMC" and "Intel PFR" in pcr 0 is given as sample. This might change once tested in real scenario.

mshamn6x committed:
added API and basic insert retrieve call for flavor template

mshamn6x committed:
added create-default-template setup task, moved API related models to pkg/hvs/model

mshamn6x committed:
setup task for default template verified

mshamn6x committed:
added source from template validation

mshamn6x committed:
intial flavor template, swagger

mshamn6x committed:
Corrected build issue and included checkmarx to pipleine

mshamn6x committed:
pipeline changes

mshamn6x committed:
Corrective formatting statements

mshamn6x committed:
Naming Conventions

mshamn6x committed:
Included all the default templates

pkumarex committed:
Update .gitlab-ci.yml

pkumarex committed:
Update .gitlab-ci.yml

pkumarex committed:
Delete flavor_schema.go

pkumarex committed:
Update flavor_controller.go

pkumarex committed:
Revert "Delete flavor_schema.go"

This reverts commit d6527d2b40625caa79e687b65da215f07d2811f9

mshamn6x committed:
Added modifictions to template

kwtj43 commented on build/ci/scan.yaml:
```diff
@@ -0,0 +1,51 @@
+services:
```
What are these changes?

ddhawal replied:
These shouldn't be part of this MR.

kwtj43 commented on go.mod:
```diff
@@ -1,5 +1,7 @@
 module github.com/intel-secl/intel-secl/v3
 
+go 1.14
```
We need to keep the go version out of go.mod.

kwtj43 commented on go.mod:
```diff
@@ -19,8 +21,9 @@ require (
 )
 
-replace github.com/vmware/govmomi => github.com/arijit8972/govmomi fix-tpm-attestation-output
+replace github.com/vmware/govmomi => github.com/arijit8972/govmomi v0.22.2-0.20200607061538-3311e9e4cdb1
```
This should also not be checked in.

kwtj43 commented on pkg/model/hvs/flavor_schema.go:
```diff
@@ -0,0 +1,32 @@
+ */
+package hvs
+
+type EventLogCreteria struct {
```
EventLogCriteria

kwtj43 commented on pkg/model/hvs/flavortemplate.go:
```diff
@@ -0,0 +1,77 @@
+
+//EventLogEquals - To store event log need be equal with specified PCR.
+type EventLogEquals struct {
+	ExculdingTags *[]string `json:"excluding_tags"`
```
I think we should avoid pointers in structures and not have pointers to arrays.  There is a need for optional json serialization (omitempty), but it also leads to a lot of nil checking or worse, the risk of panic if a nil check wasn't performed.

@ddhawal Do we have an standards around this?

mshamn6x replied:
Initially we didn't used pointer, later we noticed in response if either excluding tags or eventlog includes  are not present then response will comes like

```
"eventlog_equals":null,
"eventlog_equals":null
```
So we included pointers to avoid this responses. Likewise we have to request like this to create FT, If we remove pointers .

```
"eventlog_equals": {},
"eventlog_includes": []
```

ddhawal replied:
slice are already pointers, not sure how defining pointer to slice would help here. Handle responses in functions if they are not appropriate instead of adding a pointer to the pointer

kwtj43 commented on pkg/model/hvs/flavortemplate.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+//PCR- Tp stpre PCR index with respective PCR bank.
+type PCR struct {
```
@mshamn6x In the future you will need to use the structures defined in this package.  Ideally, it would contain all the structures and json decoration needed to share data between the Trust-agent and HVS.

kwtj43 reviewed (changes requested):
I took a first pass at the MR but didn't get to deep into the code.  @ddhawal can you take a deeper look at the postgres/router/controller code?

mshamn6x committed:
Changes to be committed:
 deleted:    build/ci/scan.yaml
 modified:   go.mod
 modified:   pkg/model/hvs/flavor_schema.go

ddhawal commented on .gitlab-ci.yml:
```diff
@@ -8,11 +8,14 @@ variables:
-  - local: '/build/ci/package-kaniko.yaml'
\ No newline at end of file
+  - local: '/build/ci/package-kaniko.yaml'
+  - local: '/build/ci/scan.yaml'
```
This file does not help running any CICD flows here, remove this from MR

ddhawal commented on build/linux/hvs/install.sh:
```diff
@@ -49,6 +50,10 @@ ln -sfT $BIN_PATH/$COMPONENT_NAME /usr/bin/$COMPONENT_NAME
 cp EndorsementCA-external.pem $CERTDIR_ENDORSEMENTCA/ && chown $SERVICE_USERNAME:$SERVICE_USERNAME $CERTDIR_ENDORSEMENTCA/EndorsementCA-external.pem
 
+# Copy Schema files
+
```
Remove unnecessary spaces

ddhawal commented on go.mod:
```diff
@@ -19,8 +21,9 @@ require (
 	github.com/spf13/viper v1.7.0
 	github.com/stretchr/testify v1.6.1
 	github.com/vmware/govmomi v0.22.2
+	github.com/xeipuuv/gojsonschema v1.2.0
```
Check if this component does not have a vulnerability

kwtj43 replied:
Harshitha ran a Snyk scan and it appears to be ok (see ISECL-12902).

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+
+	flavorTemplateReq, err := getFlavorTemplateCreateReq(r)
+	if err != nil {
+		if strings.Contains(err.Error(), "Invalid Content-Type") {
```
throw and catch custom errors instead of checking contents of the error message

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+		if strings.Contains(err.Error(), "Invalid Content-Type") {
+			return nil, http.StatusUnsupportedMediaType, &commErr.ResourceError{Message: "Invalid Content-Type"}
+		}
+		return nil, http.StatusBadRequest, &commErr.ResourceError{Message: err.Error()}
```
log errors

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	flavorTemplate, err := ftc.Store.Create(&flavorTemplateReq)
+	if err != nil {
+		defaultLog.WithError(err).Error("controllers/flavortemplate_controller:Create() Error creation flavor template")
+		return nil, http.StatusInternalServerError, &commErr.ResourceError{Message: err.Error()}
```
Use user-friendly error message instead of passing error as it is to user

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	defer defaultLog.Trace("controllers/flavortemplate_controller:Retrieve() Leaving")
+
+	templateID := uuid.MustParse(mux.Vars(r)["id"])
+	defaultLog.Debug("controllers/flavortemplate_controller:Retrieve() ID : ", templateID)
```
Is this log message required?

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	return flavorTemplate, http.StatusOK, nil
+}
+
+func validateQueryParameter(includeDeleted string) (bool, error) {
```
This function seems to be performing only boolean string validation, function name does not go with the implementation

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+		} else if strings.EqualFold(includeDeleted, "false") {
+			included = false
+		} else {
+			defaultLog.Error("controllers/flavortemplate_controller:validateQueryParameter() Invalid query parameter given")
```
log errors in calling function

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	defaultLog.Trace("controllers/flavortemplate_controller:Search() Entering")
+	defer defaultLog.Trace("controllers/flavortemplate_controller:Search() Leaving")
+
+	includeDeleted := r.URL.Query().Get("include_deleted")
```
includeDeleted is the only search parameter? shouldn't we include name, type etc

pkumarex replied:
@kwtj43 Can you please tell us what additional paramaters need to be included.

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	defer defaultLog.Trace("controllers/flavortemplate_controller:Delete() Leaving")
+
+	templateId := uuid.MustParse(mux.Vars(r)["id"])
+	defaultLog.Debug("controllers/flavortemplate_controller:Delete() ID : ", templateId)
```
Do not directly log user input, if needed to be logged, log post input sanitization or validation.

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	//Restore the request body to it's original state
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```
why this is needed?

pkumarex replied:
We need to validate the body content against schema. Once, we have decoded, we were not able to read it again.

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	defaultLog.Trace("controllers/flavortemplate_controller:getFlavorTemplateCreateReq() Entering")
+	defer defaultLog.Trace("controllers/flavortemplate_controller:getFlavorTemplateCreateReq() Leaving")
+
+	var CreateFlavorTemplateReq hvs.FlavorTemplate
```
local variables do not start with caps in Go

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+
+	var CreateFlavorTemplateReq hvs.FlavorTemplate
+	if r.Header.Get("Content-Type") != constants.HTTPMediaTypeJson {
+		secLog.Error("controllers/flavortemplate_controller:getFlavorTemplateCreateReq() Invalid Content-Type")
```
Log errors at only functions interacting with the user

mshamn6x replied:
This returns only http error which need to be returned to response, since no file and package name should be in response this cannot be logged in 'create' method so this error logged here.

ddhawal commented on pkg/hvs/controllers/flavortemplate_controller.go:
```diff
@@ -0,0 +1,263 @@
+	}
+
+	if CreateFlavorTemplateReq.ID != uuid.Nil {
+		secLog.Error("controllers/flavortemplate_controller:getFlavorTemplateCreateReq() Invalid flavor template request given ID should be NIL")
```
why the user can not give their own templates?

ddhawal commented on pkg/hvs/postgres/flavortemplate_store.go:
```diff
@@ -0,0 +1,135 @@
+	defer defaultLog.Trace("postgres/flavortemplate_store:Create() Leaving")
+
+	createdTemplate := flavorTemplate{
+		ID:      uuid.New(),
```
use uuid.NewRandom()

ddhawal commented on pkg/hvs/postgres/flavortemplate_store.go:
```diff
@@ -0,0 +1,135 @@
+			return nil, errors.Wrap(err, "postgres/flavortemplate_store:Search() - Could not scan record ")
+		}
+		//if (included && template.Deleted) || (!included && !template.Deleted) {
+		if included || (!included && !template.Deleted) {
```
What are we trying to achieve here? what does included mean?

ddhawal commented on pkg/hvs/postgres/flavortemplate_store.go:
```diff
@@ -0,0 +1,135 @@
+		if err := rows.Scan(&template.ID, (*PGFlavorTemplateContent)(&template.Content), &template.Deleted); err != nil {
+			return nil, errors.Wrap(err, "postgres/flavortemplate_store:Search() - Could not scan record ")
+		}
+		//if (included && template.Deleted) || (!included && !template.Deleted) {
```
remove commented code

ddhawal commented on pkg/hvs/tasks/create_default_template.go:
```diff
@@ -0,0 +1,464 @@
+	"default-pfr",
+}
+
+// var defaultFlavorTemplateNames = []string{
```
remove commented code

ddhawal commented on pkg/hvs/tasks/create_default_template.go:
```diff
@@ -0,0 +1,464 @@
+	}
+
+	for _, ft := range templates {
+		// create default flavortemplates ONLY if it does not exist already
```
how this is achieved?

ddhawal commented on pkg/hvs/tasks/create_default_template.go:
```diff
@@ -0,0 +1,464 @@
+		defaultFlavorTemplateMap[templateName] = false
+	}
+
+	ftList, err = ftStore.Search(false)
```
this is not an effective search, search by template names would be a better implementation. The implemented method would have a performance impact.

mshamn6x replied:
since search is implemented with only whether to include deleted templates or not, this implemented in this way. Once kent provided comment on this, we can modify with name search.

ddhawal commented on pkg/lib/flavor/types/linux_platform_flavor.go:
```diff
@@ -14,7 +16,6 @@ import (
 	hcTypes "github.com/intel-secl/intel-secl/v3/pkg/lib/host-connector/types"
 	taModel "github.com/intel-secl/intel-secl/v3/pkg/model/ta"
 	"github.com/pkg/errors"
-	"strings"
```
this file can be ignored

ddhawal commented on pkg/model/hvs/flavortemplate.go:
```diff
@@ -0,0 +1,77 @@
+
+//EventLogEquals - To store event log need be equal with specified PCR.
+type EventLogEquals struct {
+	ExculdingTags *[]string `json:"excluding_tags"`
```
ExcludingTags

ddhawal commented on pkg/model/hvs/flavortemplate.go:
```diff
@@ -0,0 +1,77 @@
+	Pcr              PCR             `json:"pcr"`
+	PcrMatches       bool            `json:"pcr_matches"`
+	EventlogEquals   *EventLogEquals `json:"eventlog_equals,omitempty"`
+	EventlogIncludes *[]string       `json:"eventlog_includes"`
```
same comment as previous here

ddhawal commented on pkg/model/hvs/flavortemplate.go:
```diff
@@ -0,0 +1,77 @@
+	ID          uuid.UUID   `json:"id" gorm:"primary_key;type:uuid"`
+	Label       string      `json:"label"`
+	Condition   []string    `json:"condition" sql:"type:text[]"`
+	FlavorParts FlavorParts `json:"flavor_parts,omitempty" sql:"type:JSONB"`
```
FlavorParts - should be pointer

ddhawal commented on pkg/hvs/postgres/models.go:
```diff
@@ -67,6 +70,13 @@ type (
 	}
 
+	//FlavorTemplate - To maintain all values keep together inorder to maintain flavor template.
+	flavorTemplate struct {
```
add a new table to migrate

pkumarex replied:
Added the table.

mshamn6x committed:
Address review comments in controller and store

mshamn6x committed:
Removed line break in install.sh

mshamn6x committed:
Changes to be committed:
modified:   pkg/hvs/controllers/flavortemplate_controller.go
modified:   pkg/hvs/tasks/create_default_template.go
modified:   pkg/model/hvs/flavortemplate.go

mshamn6x committed:
get latest source changes and merge

mshamn6x committed:
changed gitlab-ci.yml and linux_platform_flavor.go

mshamn6x committed:
Add type assertions for errors

mshamn6x committed:
modified flavortemplate_store

mshamn6x committed:
resolved user type error
